src/server/replica.cc

```diff
@@ -705,14 +707,61 @@ void Replica::StableSyncDflyFb(Context* cntx) {
       cntx->Error(res.error(), "Journal format error");
       return;
     }
-
-    executor.Execute(std::move(res.value()));
-
+    ExecuteEntry(&executor, std::move(res.value()));
     last_io_time_ = sock_->proactor()->GetMonotonicTimeNs();
   }
   return;
 }
 
+void Replica::ExecuteEntry(JournalExecutor* executor, journal::ParsedEntry&& entry) {
+  if (entry.shard_cnt <= 1) {  // not multi shard cmd
+    executor->Execute(std::move(entry));
+    return;
+  }
+
+  // Multi shard command flow:
+  // step 1: Fiber wait untill all the fibers that should execute this tranaction got
+  // to the journal entry of the transaction.
+  // step 2: execute the command (All fibers)
+  // step 3: Fiber wait untill all fibers finished the execution
+  // By step 1 we enforce that replica will execute multi shard commands that finished on master
+  // Step 3 ensures the correctness of flushall/flushdb commands
+  // TODO: this implemantaion does not support atomicity in replica
+  // Although multi shard transaction happen in step 2 very close to each other,
+  // user can query replica between executions.
+  // To support atomicity we should have one fiber in step 2 which will excute all the entries of
+  // the transaction together. In case of global comand such as flushdb the command can be executed
+  // by only one fiber.
+  // TODO: support error handler in this flow
+
+  // Only the first fiber to reach the transaction will create data for transaction in map
+  multi_shard_exe_->map_mu.lock();
+  auto [it, was_insert] = multi_shard_exe_->tx_sync_execution.emplace(entry.txid, entry.shard_cnt);
+
+  // Note: we must release the mutex befor calling wait on berrier
+  multi_shard_exe_->map_mu.unlock();
+
+  VLOG(2) << "txid: " << entry.txid << " unique_shard_cnt_: " << entry.shard_cnt
+          << " was_insert: " << was_insert;
+
+  // step 1
+  it->second.berrier.wait();
+  // step 2
+  executor->Execute(std::move(entry));
+  // step 3
+  it->second.berrier.wait();
+
+  // Note: erase from map can be done only after all fibers returned from wait.
+  // The last fiber which will decrease the counter to 0 will be the one to erase the data from map
+  auto val = --it->second.counter;
+  VLOG(2) << "txid: " << entry.txid << " unique_shard_cnt_: " << entry.shard_cnt
```

Comment:
1. since you move the entry above, reading its fields can technically show wrong values. 
2. You do not use `was_inserted` to check that only a single fiber called flushdb. why?
3. Please use `fetch_sub` when decrementing atomics. you will need to compare `val` with 1.

Reply:
1. We can relax the interface on Execution to accept a mut-ref. We need mut because of the mutable CmdArgList in Dispatch (which can be changed with ToUpper etc inside the command)
2. Because we can't tell now what kind of command that is - sharded part or global. We can, theoretically, check with the number of arguments it has or pass additional info. For now, the other executions on FLUSHDB will be no-ops (but redundant transactions)

Reply:
Regarding number 2, I will but not in this PR.
There are 2 TODO notes which I added.
I will work on another PR to support the execution of the commands by a single fiber, which will require different flow for global commands and non global commands.
I will work on another PR to support the cancellation flow, once the PR I created for heilo with barrier will be merged